### PR TITLE
Bug 2008312: Add e2e test for Active Health Checks

### PR DESCRIPTION
**Pull request by rexagod**
Added an e2e test for the recently introduced Active Health Checks
feature for when there's a warning or an error. Also, I noticed some
wrapping, that was happening for a few lengthy messages (like the
one below) and set the `flexWrap` to `nowrap` inorder to avoid that
behaviour, even when the troubleshoot link is displayed.

2021-08-02-042434_365x261_scrot

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:9]: This should be selected via data-test ids. Using url links is not long term solution. Can you take a look ?

Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:9]: Also create a separate method to visit these links

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:41]:
```suggestion
    const restoreResources = scaleDownAndOpenHC(State.Warning, [Deployments.ROOK_MON_A, Deployments.ROOK_MGR_A]);
```
Enumize these hardcoded strings.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:18]: Create a separate helper method `scaleDeployment(deployment: string, replicas: number)` for this ? Check if we already have one

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:42]: Same here
Create an object for the possible messages. Seems like a place for human errors.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:27]:
```suggestion
  deployments.forEach(deployment =>  scaleDeployment(deployment, replicas));
```
Very odd pattern used here

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:48]: Place these in the above object.
Have a very well grouped object for messages 
```
msgs = {
errors : {
....
....
....
}
warning: {
....
....
}
}
```

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:49]: You can move these scaling operations to this test itself
```
scaleDeps(....)
verifyMessages(.....)
scaleDeps(....)
```

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:24]:
```suggestion
    cy.byTestID('storage-cluster-health-item')
```
Odd way to name the data-tests. Can you change all of them to the pattern we usually follow ?

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:54]:
```suggestion
    cy.byTestID('storage-cluster-health-item-warning-secondary-status', { timeout: 2 * MINUTE }).should('not.exist');
```
Name can be more specific

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:55]:
```suggestion
    cy.byTestID('storage-cluster-health-item-error-secondary-status', { timeout: 2 * MINUTE }).should('not.exist');
```

Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:55]: by the way why is it secondary-status ? is there a primary status too ?

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:52]: Make this the first test so that it complies with the test pattern we established above

**Review by vbnrh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:27]: I've used a [closure pattern (lexical scoping)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#lexical_scoping) to save the state of `scaleTo` when the argument is `1` so that I could easily invoke that from the calling parent function on demand.

https://github.com/openshift/console/pull/9690/files#diff-fa20d6bd8341e9a45df0d338a1cee9a6490b5f86dfe10f011c4c6e03b61c63b2R43

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:18]: > Create a separate helper method scaleDeployment(deployment: string, replicas: number) for this ?

Sure.

> Check if we already have one

I looked into it. It seems we don't. See https://github.com/rexagod/console/blob/a462dbee4868dbfb5e94a92edcba0dad7433cf8d/frontend/packages/ceph-storage-plugin/integration-tests/tests/2-tests/ocp-dashboard-card-healthcheck.scenario.ts#L35.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:49]: Sure, will do.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:55]: I named it after the [component](https://github.com/rexagod/console/blob/a462dbee4868dbfb5e94a92edcba0dad7433cf8d/frontend/packages/console-shared/src/components/status/SecondaryStatus.tsx#L9) where it was defined. This is basically the status that shows below the "Storage Cluster" title on the OCS Dashboard.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:27]: ok, i get it but i don't seem to see why we need this additional complexity here. We can just scale down the pods here itself or do it in the tests scope, it will be much better

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (rexagod) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:17]: I've kept this to ERROR as `error` is what we get from the JSON response from the `ceph` object and so that it complies with [this](https://github.com/openshift/console/pull/9690/files#diff-fa20d6bd8341e9a45df0d338a1cee9a6490b5f86dfe10f011c4c6e03b61c63b2R43). Please refer the [source](https://github.com/openshift/console/blob/110499002d84fe9a24c2533c96edd86e4466029c/frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx#L106) for how we catch these (notice that we have only a single straightforward `error` in the RegEx).

**Comment by rexagod** [comment]
/test ceph-storage-plugin
Cluster creation failed.

**Comment by rexagod** [comment]
/test ceph-storage-plugin
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9690/pull-ci-openshift-console-master-ceph-storage-plugin/1427529032286605312#1:build-log.txt%3A45

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx
Comment (bipuladh) [inline frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:55]: This a bug fix.

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx
Comment (bipuladh) [inline frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:24]: This is only applicable to us. So we should make changes to title before passing here. Thank making this a generic logic.

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:60]: This should be a bug fix and not a test.

**Review by bipuladh — CHANGES_REQUESTED** [review]

**Comment by rexagod** [comment]
/hold
Until bug fix PR is merged.

**Comment by rexagod** [comment]
/test ceph-storage-plugin
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9690/pull-ci-openshift-console-master-ceph-storage-plugin/1427728759720513536#1:build-log.txt%3A44

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:68]: Let's not perform these replacements.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/active-health-checks.spec.ts:52]: Move them out of the `describe` block.

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx
Comment (rexagod) [reply frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx]: Thank you, Sam. I sent a fixup commit to verify this and this was indeed the problem.

**Comment by rexagod** [comment]
Closing this for now: https://bugzilla.redhat.com/show_bug.cgi?id=2008312#c1. I'll migrate this to ODF Console in a different PR as the CI is up and running there now.

Thanks.

**Pull request closed without merging**
